design/thin-edge-core.md

```diff
@@ -0,0 +1,50 @@
+# Thin-edge Core Vision
+
+A typical IoT application has to address:
+
+* __sophisticated business-cases__, with applications involving several vendors, actors and users;
+* __very diverse integration needs__ over operating systems, protocols, data sources, processing tools, actuators, cloud end-points;
+* __constraint environments__ with low-resources devices, large-scale deployments, high exposure to security risks,
+  prohibitive manual operations, long-term supports;
+* __safety related requirements__ with devices operating machines in hazardous contexts.
+  
+The forces being these requirements are pushing in opposite directions.
+
+* On one side, one needs a dynamic and open setting with a large set of features that can be easily extended and combined.
+  [NodeRed](https://nodered.org/) is the archetype of such tools facilitating connections of components.
+* On the other side, one needs a minimal system specifically built for a task with cherry-picked components.
+  On the extreme side this leads to the idea of [unikernel](https://en.wikipedia.org/wiki/Unikernel) and sealed, fixed-purpose software images.
+  
+Can these two poles be reconciled?
+
+The approach of thin-edge is to ease the development of IoT applications on edge devices with a smooth transition between:
+* prototyping use-cases - when the aim is to easily explore and combine features,
+* production use-cases - when the need is to deploy a hardened specific software on a large fleet of devices,
+* as well as intermediate use-cases - when the application is build by several vendors with components written using different programming languages;
+  or when a manufacturer assembles his tools on the devices while letting open to his customers the option to add their own modules. 
+
+To achieve this goal, the foundations of Thin-edge address the hardened case first.
+* The core of thin-edge and its components are written using the *Rust* programming languages.
+* The set of components used by an application is defined at *build time*.
+* These components cooperate using a streaming, asynchronous message passing, *internal* API.
+* Connections to the outside are delegated to specific bridge components
+  that abstract the actual protocol (MQTT, HTTP, Modbus, OPC UA)
+  while making these external channels accessible to the other components.
+* The *internal* messages exchanged by the components are *predefined* and cover the domain of telemetry and IoT
+  (measurements, events, alarms, operation requests, operation outcomes ...). 
+  Messages sent to the outside are freely defined by the respective bridge components.
+* The core of thin-edge provides the tools to configure the components as well as the internal message routes.
+* An executable for an IoT application is statically defined by an assemblage of components and their configuration.
```

Comment:
What do you see in that context exactly as the scope of that "IoT application"? What is part of that application?
1) thin-edge platform (i.E. the core)?
2) plugin and external added components?
3) customer specific configurations?
4) customer specific application with business logic?
5) more?

Reply:
First, it's important to differentiate the IoT application from the executables. The IoT application encompasses not only the devices but also the cloud and other systems. Hence, I take your question as related to thin-edge *executables* running on a device.

So what is included in a thin-edge executable:

1. The core engine (the plugin API and the mechanism to build, connect and run plugins).
2. Any Rust/internal plugin builder that has been explicitly added as a dependency of the executable.
3. The associated configuration has to provide a configuration entry for each of the plugin instances to be created.
4. Customer specific application logic can be implemented as a plugin and included into the executable (at compile time).
5. Customer specific application logic don't have to be included in such an executable - but for that to work the executable must include a bridge/connector component for the protocol used by the customer specific process.

For instance, to implement somethings akin to the current c8y mapper, one needs to assemble these plugins:
* An MQTT connector plugins that defines the misc thin-edge JSON payload and topics
* A c8y plugin that is roughly what we name currently the c8y convertor.
* The configuration will be mostly related to the c8y end-point.

To implement a tedge_agent, one needs:
* The MQTT connector plugin
* An sm management plugin that handles software-management
* A rust plugin connector used to connect an external sm plugin via the command line
* The configuration gives the location of the external sm-plugins.

A really nice benefit of this architecture is that the *same* plugins can be arranged into an executable tuned for a specific use case. For instance, the agent and the mapper can be combine into a single executable:
* The MQTT connector plugin
* The c8y plugin
* The sm management plugin
* The sm cli plugin connector
* The configuration gives the location of the external sm-plugins and of c8y.

And for a very specific and tuned use-case, say no MQTT external clients and only APT packages. The same can be achieved with a few adjustements:

* A plugin for direct MQTT connection to the cloud (no thin-edge json, no mosquitto).
* The c8y plugin
* The sm management plugin
* A rust implementation of the APT sm plugin.
* The configuration gives the location of c8y.